packages/url_launcher/lib/src/app_control.dart

```diff
@@ -2,158 +2,104 @@
+    _appControlSetUri = libAppControl
+        .lookup<NativeFunction<AppControlSetUriFunc>>('app_control_set_uri')
         .asFunction();
-    _sendLaunchRequest = lib
-        .lookup<NativeFunction<_app_control_send_launch_request>>(
+    _appControlSendLaunchRequest = libAppControl
+        .lookup<NativeFunction<AppControlSendLaunchRequestFunc>>(
             'app_control_send_launch_request')
         .asFunction();
-    _destroy = lib
-        .lookup<NativeFunction<_app_control_destroy>>('app_control_destroy')
+    _appControlDestroy = libAppControl
+        .lookup<NativeFunction<AppControlDestroyFunc>>('app_control_destroy')
+        .asFunction();
+
+    final DynamicLibrary libBaseCommon =
+        DynamicLibrary.open('libcapi-base-common.so.0');
+    _getErrorMessage = libBaseCommon
+        .lookup<NativeFunction<GetErrorMessageFunc>>('get_error_message')
         .asFunction();
   }
 
-  late int Function(Pointer<Pointer<_AppControl>>) _create;
-  late int Function(Pointer<_AppControl>, Pointer<Utf8>) _setOperation;
-  late int Function(Pointer<_AppControl>, Pointer<Utf8>) _setUri;
-  late int Function(Pointer<_AppControl>, Pointer<Void>, Pointer<Void>)
-      _sendLaunchRequest;
-  late int Function(Pointer<_AppControl>) _destroy;
+  late AppControlCreate _appControlCreate;
+  late AppControlSetOperation _appControlSetOperation;
+  late AppControlSetUri _appControlSetUri;
+  late AppControlSendLaunchRequest _appControlSendLaunchRequest;
+  late AppControlDestroy _appControlDestroy;
+  late GetErrorMessage _getErrorMessage;
 
-  Pointer<_AppControl> _handle = nullptr;
-
-  /// Corresponds to `app_control_create()`.
-  void create() {
-    final Pointer<Pointer<_AppControl>> pHandle = malloc();
-    final int ret = _create(pHandle);
-    _handle = pHandle.value;
-    malloc.free(pHandle);
+  AppControlHandle _handle = nullptr;
 
+  void _throwOnError(int ret) {
     if (ret != 0) {
+      final Pointer<Utf8> message = _getErrorMessage(ret);
       throw PlatformException(
-        code: _getErrorCode(ret),
-        message: 'Failed to execute app_control_create.',
+        code: ret.toString(),
+        message: message.toDartString(),
       );
     }
   }
 
+  /// Corresponds to `app_control_create()`.
+  void create() {
+    final Pointer<AppControlHandle> pHandle = malloc();
+    final int ret = _appControlCreate(pHandle);
+    _handle = pHandle.value;
+    malloc.free(pHandle);
+    _throwOnError(ret);
+  }
+
   /// Corresponds to `app_control_set_operation()`.
   void setOperation(String operation) {
     assert(_handle != nullptr);
-
-    final int ret = _setOperation(_handle, operation.toNativeUtf8());
-    if (ret != 0) {
-      throw PlatformException(
-        code: _getErrorCode(ret),
-        message: 'Failed to execute app_control_set_operation.',
-      );
-    }
+    using((Arena arena) {
+      _throwOnError(_appControlSetOperation(_handle, operation.toNativeUtf8()));
```

Comment:
```suggestion
      _throwOnError(_appControlSetOperation(_handle, operation.toNativeUtf8(arena)));
```
Maybe you should pass `arena` as the allocator to `toNativeUtf8()` method. I am not sure. please check one more.